tools/testing/gatekeeper/jobs.py

```diff
@@ -0,0 +1,189 @@
+            self.results[job_name] = RUNNING
+            return
+        if job["conclusion"] != "success":
+            self.results[job_name] = job['conclusion']
+            return
+        self.results[job_name] = PASS
+
+    def status(self):
+        """
+        :returns: 0 - all tests passing; 127 - no failures but some
+            tests in progress; 1 - any failure
+        """
+        running = False
+        if not self.results:
+            # No results reported so far
+            return FAIL
+        for status in self.results.values():
+            if status == RUNNING:
+                running |= True
+            elif status != PASS:
+                # Status not passed
+                return FAIL
+        if running:
+            return RUNNING
+        return PASS
+
+    def __str__(self):
+        """Sumarize the current status"""
+        good = []
+        bad = []
+        warn = []
+        for job, status in self.results.items():
+            if status == RUNNING:
+                warn.append(f"WARN: {job} - Still running")
+            elif status == PASS:
+                good.append(f"PASS: {job} - success")
+            else:
+                bad.append(f"FAIL: {job} - Not passed - {status}")
+        out = '\n'.join(sorted(good) + sorted(warn) + sorted(bad))
+        stat = self.status()
+        if stat == RUNNING:
+            status = "Some jobs are still running."
+        elif stat == PASS:
+            status = "All required jobs passed"
+        elif not self.results:
+            status = ("No required jobs for regexps: " +
+                      ";".join([_.pattern for _ in self.required_regexps]))
+        else:
+            status = "Not all required jobs passed!"
+        return f"{out}\n\n{status}"
+
+    def get_jobs_for_workflow_run(self, run_id):
+        """Get jobs from a workflow id"""
+        total_count = -1
+        jobs = []
+        page = 1
+        while True:
+            response = requests.get(
+                f"{_GH_RUNS_URL}/{run_id}/jobs?per_page=100&page={page}",
+                headers=_GH_HEADERS,
+                timeout=60
+            )
+            response.raise_for_status()
+            output = response.json()
+            jobs.extend(output["jobs"])
+            total_count = max(total_count, output["total_count"])
+            if len(jobs) >= total_count:
+                break
+            page += 1
+        return jobs
+
+    def check_workflow_runs_status(self):
+        """
+        Checks if all required jobs passed
+
+        :returns: 0 - all passing; 1 - any failure; 127 some jobs running
+        """
+        # TODO: Check if we need pagination here as well
+        latest_commit_sha = os.getenv("COMMIT_HASH")
+        response = requests.get(
+            _GH_RUNS_URL,
+            params={"head_sha": latest_commit_sha},
+            headers=_GH_HEADERS,
+            timeout=60
+        )
+        response.raise_for_status()
```

Comment:
In two places it can crash this script (` response.raise_for_status()` throwing exception due API unavailable), failing the gatekeeper job and consequently blocking a PR. Can we resume the checking by re-running the gatekeeper job from the Github web UI?

Regardless, I think we should implement a retry mechanism to call the Github API. The interwebs says `requests` has a retry implementation with backoff!

Reply:
At this point I'd like to keep it this way to experiment with the limits. If it becomes an issue, we should get the exception with details. I haven't found the limits for jobs on GH runners but somewhere I read there should not be any limit.

Anyway it's just another workflow, one can re-try it from the UI.

Reply:
That's a good idea. Let's see how it works in production then we fine tune.